vmware-tanzu/nsx-operator
# Support podSelector namespaceSelector conversion for SP in VPC mode

This patch is to:
1. Support changing SecurityPolicy Rule peer(source/destination) group
from podSelector to namespaceSelector in VPC mode, and vice versa.
2. Remove duplicated project shares and project groups when expanding rules.
3. Fix the bug that rule store is not cleaned in security policy GC or
cleanup process.
4. Split project group store from group store which is for storing groups
in VPC level.
5. Wrap project groups and shares with SecurityPolicy, groups, rules in
VPC level into one hierarchy resource tree for OrgRootClient API call.
So, there is only one time API call for SecurityPolicy
create/update/delete.
6. Enhance log output to use one word for log info as the key index.

Test Done:
1. Create a SecurityPolicy, and then apply a change from podSelector to namespaceSelector in VPC mode.
2. Create a SecurityPolicy, and then apply a change from namespaceSelector to podSelector in VPC mode.
3. Create a SecurityPolicy, and then delete a rule with namespaceSelector.
4. Create a SecurityPolicy, and then delete a port entry from the rule with namespaceSelector.
5. Create a SecurityPolicy, and  then delete a rule with namespaceSelector and add a rule with namespaceSelector
at the same time.
7.  Create a SecurityPolicy with namespaceSelector and podSelector in non-VPC mode with v4.1.2 branch, 
and then restart NSX operator with patch change to see if operator upgrade from v4.1.2 to current 
patch version is ok. 
8. Create/update/delete a SecurityPolicy with namedport.
9. Create/update/delete a NetworkPolicy with namedport.
8. GC in VPC mode.
10. GC in non-VPC mode.
